\documentclass{article}
\usepackage{amsfonts}
\usepackage[cm]{fullpage}
\usepackage{graphicx}
\usepackage{hyperref}
\graphicspath{
{/home/jburdick/gcb/work/svnroot/trunk/src/}
{/home/jburdick/gcb/}
}
\begin{document}

\subsection*{Cell sorting matrix}

The unmixing methods require an estimate of which cell is in which sorted fraction,
and how much RNA each cell contributes to the sample.

Let $S_{ij}$ be
the probability that a given cell was present in a given sorted fraction. Let $V_j$
be the volume of the $j$th cell. Then we define

\[
M_{ij} = \frac{S_{ij} V_j}{ \sum\limits_{i}^{} S_{ij} V_j }
\]

\subsubsection*{Including temporal information}

We attempted to increase the temporal resolution of our predictions
by including RNA-seq data from precisely-staged embryos.
This included 23 timepoints spaced between zero and six hours after hatching.
We only included the timepoints up until seven hours, which is approximately
when enclosure occurs.

These embryos sequenced in that experiment were grown at $20^{\circ}$C.
Since our image data were aligned to the lineage of an embryo which was
grown at the same temperature, the data sets theoretically should be fairly
well aligned.
We did not attempt to align the timepoints more precisely with our
image data. However, adjacent timepoints are fairly highly correlated.

\subsubsection*{Image thresholding}

We estimated $S_{ij}$ based on 29 movies.
For each movie, we manually chose an intensity distribution for cells expressing
or not expressing the reporter, and 
used this logistic model to classify each cell as ``expressing'' or ``not expressing.''
We assumed that the probability of a cell being in a negative fraction was
one minus the probability of a cell being in a positive fraction.

\subsubsection*{Estimating cell volume}

We assumed that the amount of RNA in each cell was proportional to the cell's volume,
times the number of time steps that the cell existed.
We estimated cell volume by assuming that each division was exactly equal, and so

\[
V_j = 2^{-\mathrm{branchlength}}
\]

This is an approximation which seems somewhat unrealistic: for instance, it predicts
that Z3 (which has undergone five divisions) has sixteen times the volume of 
ABalaaaala (which has undergone nine divisions.)

\subsubsection*{Estimating biological noise between measurements}

We estimated biological noise between replicates using four samples which were
grown and sorted independently. For each (on log scale), we plotted the variance
in read depth between replicates, as a function of average read depth.

\includegraphics[width=0.7\textwidth]{git/unmix/seq/quant/noiseBetweenReplicates.pdf}

The average slope among these experiments (assuming a zero $y$-intercept) was 1.05.
Therefore (for the sake of simplicity)
we assumed that the variance equalled the number of reads.

\subsubsection*{Sorting purity}

The sorting purities (measured by re-sorting the sorted cells) ranged from 82\% to 97\%.
This means that the measured depletion of a gene in a given sorted fraction is less than
the actual depletion. (This is important, because we're using depletion of a gene in a
sorted fraction to ``rule out'' expression in some cells.)

Let $S$ be the measured expression of a gene in the singlet cells,
$F$ be the
measured expression of a gene in a sorted fraction, with known sort purity $p$, and
$G$ be the true expression of that gene in the sorted fraction.
We can model $F$ as a mixture of $S$ and $G$:

\[
F = pG + (1-p)S
\]

If we assume that $S$ and $F$ are normally distributed (with variance estimated as described
above), then we can obtain $G$'s distribution, which is also normal.

Our estimate of $G$ will be smaller than our estimate of $F$.
Indeed, this can result in estimates of $G$ which are negative; in such
cases, we set the mean to zero.

In cases where the sort purity wasn't measured, we assumed it was 100\%.
This means we will underestimate the depletion of a gene in a fraction,
and so presumably our estimate will be less tissue-specific.

\newpage

\subsection*{Methods of unmixing}

We tried several methods of estimating the expression in each cell, based
on the expression in fractions.

Not all 1,341 cells present at hatch were included in our measurements,
which are unlikely to include cells after ventral enclosure (around 400 minutes.)
We assumed enclosure happened at 420 minutes, including all cells born
before that time except P0; this includes 1,312 cells.

If a gene's expression in a fraction was zero, we assumed that the expression
of that gene in all cells in that fraction was zero.

\subsection*{Using the pseudoinverse}

\[
x = M^\dagger G
\]

This may result in negative predictions; we replaced any negative elements
of $x$ with 0.

\subsubsection*{Using the pseudoinverse, constraining $x \ge 0$}



\subsection*{Unmixing using EP}





Convergence of EP can be problematic. We 
used damping (with $\alpha = 0.5$) to improve convergence.
We used a prior that was 1,000 times the maximum estimated variance
in any fraction; using a flat
prior caused numerical problems.


\end{document}

