\documentclass[serif,9pt]{beamer}
\usepackage[latin1]{inputenc}
\usepackage{multirow}
\renewcommand{\sfdefault}{times}

% \usetheme{Warsaw}
\AtBeginSection[]
{
  \begin{frame}{Outline}
 %   \frametitle{\thesection}
    \tableofcontents[currentsection]
  \end{frame}
}
\AtBeginSubsection[]
{
    \begin{frame}{Outline}
        \tableofcontents[currentsection,currentsubsection]
    \end{frame}
}

\graphicspath{
{/home/jburdick/gcb/work/svnroot/trunk/tex/image/}
{/home/jburdick/gcb/work/svnroot/trunk/tex/thesis/proposal/image/}
{/home/jburdick/gcb/work/svnroot/trunk/src/}
{/home/jburdick/gcb/}
}
\beamertemplatenavigationsymbolsempty

\title{Where are genes expressed in the {\em C. elegans} embryo?}
\author{Josh Burdick \\ Murray Lab}
\date{September 13, 2012}
\begin{document}

\begin{frame}
\titlepage
\end{frame}

\begin{frame}
\tableofcontents
\end{frame}

\section{Introduction}

\begin{frame}{Where a gene is expressed is important}

\begin{minipage}{0.27\textwidth}
\includegraphics[width=\textwidth]{pha-4_emb_expr.png}

\hfill {\footnotesize Mango {\em et al} 1994}
\end{minipage}
\pause
\hspace{0.01\textwidth}
\begin{minipage}{0.7\textwidth}
\includegraphics[width=\textwidth]{pha-4_phenotype.png}

\hfill {\footnotesize Kalb {\em et al} 1998}
\end{minipage}

\end{frame}

\begin{frame}{Where {\em many} genes are expressed is important}

\includegraphics[width=0.9\textwidth]{yanai2008.png}

\hfill Yanai, 2008

\end{frame}

\begin{frame}{Measuring expression using automated lineage tracing}

\includegraphics[width=\textwidth]{DeepLineageTracing.png}

\hfill Murray {\em et al}, 2012

\end{frame}

\begin{frame}{Measuring expression using automated lineage tracing}

{\center
\includegraphics[width=0.75\textwidth]{image/movie/P0_1200x400/20110929_ceh-43_L1.png}

\includegraphics[width=0.75\textwidth]{image/movie/P0_1200x400/20101213_RW10889_pal-1_L3.png}

\includegraphics[width=0.75\textwidth]{image/movie/P0_1200x400/20110531_pha4_24C_L3.png}

}

\end{frame}

\section{Sequencing of 14 flow-sorted samples}

\begin{frame}{Samples sorted}

\includegraphics[width=\textwidth]{FACSlog.pdf}

\hfill Travis Walton

\begin{itemize}

\item promoter fusions are mCherry, protein fusions are GFP

\item additional controls (for {\em cnd-1} and {\em pha-4}):

\begin{itemize}

\item ungated: cells which were dissociated, but not gated

\item singlets: cells which were gated by size, to probably not be clumps

\end{itemize}

\end{itemize}

\end{frame}

\subsection{Reproducibility}

\begin{frame}{How reproducible are the data?}

\includegraphics[width=0.8\textwidth]{git/unmix/seq/quant/replicateScatterPlot.png}

\end{frame}

\begin{frame}{Which samples are similar to each other?}

\includegraphics[width=1\textwidth]{git/unmix/seq/cluster/pvclust.pdf}

\end{frame}

\subsection{Cells included in sorting}

\begin{frame}{Are some cells depleted from the sorting process?}

\begin{minipage}{0.58\textwidth}
\includegraphics[width=\textwidth]{git/unmix/seq/quant/scatterplot/cnd-1_ungated.png}
\end{minipage}
\pause
\begin{minipage}{0.4\textwidth}
Tissues depleted from singlets
\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
intestine & 2234 & $10^{-34}$ \\
germ line & 295 & $10^{-32}$ \\
hypodermis & 1036 & $10^{-16}$ \\
\end{tabular}
\end{table}
\end{minipage}

\end{frame}

\subsection{Times included by sorting}

\begin{frame}{Some genes are mostly expressed at a particular time}

\includegraphics[height=0.8\textheight]{git/unmix/seq/timing/gene_timeseries.pdf}

\hfill Boeck and Waterston (modENCODE)

\end{frame}

\begin{frame}{What times are included in the sorting?}

\begin{centering}
\includegraphics[width=\textwidth]{git/unmix/seq/timing/perStageMarkersControls.pdf}
\end{centering}

\end{frame}

\begin{frame}{Sequencing of 14 flow-sorted samples}

\begin{itemize}

\item Data are fairly reproducible between biological replicates.

\item Gating for single cells depletes some tissues, particularly
intestine and hypodermis.

\item Gating for single cells enriches somewhat for later embryonic
cells (after 200 minutes.)

\end{itemize}

\end{frame}

\section{Transcriptomes of cells expressing a marker}

\subsection{Genes enriched in fractions}

\begin{frame}{{\em ceh-26}}

\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{git/unmix/seq/quant/scatterplot/ceh-26.png}
\end{minipage}
\begin{minipage}{0.58\textwidth}

{\small Expressed in excretory cell, some pharyngeal gland cells}

\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
excretory secretory system & 5 & 0.006 \\
\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline
integral to membrane & {\em srx-47, vha-17, clc-2} (23 others) & $10^{-6}$ \\
astacin activity & {\em nas-12, nas-2, Y19D10A.6, nas-31, toh-1} & 0.017 \\
carbohydrate binding & {\em clec-185, clec-65, clec101} (+ 10 others) & 0.04 \\
\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20120401_JIM122_L3.png}

\end{frame}

\begin{frame}{{\em hlh-16}}

\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{git/unmix/seq/quant/scatterplot/hlh-16.png}
\end{minipage}
\begin{minipage}{0.58\textwidth}
{\small Expressed in excretory duct and pore,
neurons (AWC, SAAVL, AIYL, SMDDL, SIAD, DB1)}

\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
phasmid neuron & 168 & $10^{-9}$ \\
amphid neuron & 271 & $10^{-5}$ \\
AWCR/L & 83 & $10^{-4}$ \\
VB neuron & 58 & $10^{-4}$ \\
DB neuron & 67 & $10^{-3}$ \\
ASHR/L & 91 & $10^{-3}$ \\
\hline

\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline
signal transducer activity & {\em gpa-13, acr-15, srg-30} (+ 66 others) & $10^{-17}$ \\
neurotransmitter binding & {\em npr-9, lgc-31, flp-9} (+ 14 others) & $10^{-8}$ \\
\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20111011_L1.png}

\end{frame}

\begin{frame}{{\em ceh-6}}

\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{git/unmix/seq/quant/scatterplot/ceh-6.png}
\end{minipage}
\begin{minipage}{0.58\textwidth}

{\small Expressed in excretory duct and socket, rectal epithelium,
neurons (DB, SMBVR, PVT, RIS),
two pharyngeal gland cells (g2L/R) }

\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
ventral cord neuron & 607 & $10^{-10}$ \\
nerve ring & 632 & $10^{-8}$ \\
preanal ganglion & 25 & 0.004 \\
\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline
G-protein coupled receptor signalling & {\em srw-113, gnrr-7, srg-31} (12 total) & $10^{-4}$ \\
axon guidance & {\em ceh-17, zag-1, unc-6} & $10^{-3}$ \\
\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20120331_RW10871_L4.png}

\end{frame}

\begin{frame}{{\em cnd-1}}
\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{git/unmix/seq/quant/scatterplot/cnd-1_12_14.png}
\end{minipage}
\begin{minipage}{0.58\textwidth}

{\small Expressed in various AB neurons, pharyngeal gland cell (g2L)}

\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
Q(L/R) {\tiny (ABp(l/r)apapaaa)} & 8 & 0.003 \\
Z1.aa {\tiny (anterior distal tip cell)} & 3 & 0.009 \\
C(a/p)p & 25 & 0.02 \\
\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline
regulation of transcription & {\em ceh-5, cnd-1, Y51H4A.4, unc-3, hlh-32} & 0.02 \\
oviposition & {\em cnd-1, bar-1, ZK930.3} & 0.036 \\
\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20110731_RW10434_L1.png}

\end{frame}

\begin{frame}{Sorting enriches sorted fractions at times when reporters are on}

\begin{centering}
\includegraphics[width=\textwidth]{git/unmix/seq/timing/perStageMarkersVsSinglets.pdf}
\end{centering}

\end{frame}

\begin{frame}{Genes enriched in sorted fractions}

\begin{table}\tiny
\begin{tabular}{lll p{1.2in} p{1.2in} }
sort & \# genes & \# genes & enriched GO & enriched \\
marker & enriched & depleted & categories & tissues \\
\hline
{\em ceh-26} & 515 & 2316 & membrane, astacin activity & excretory secretory system \\
{\em ceh-27} & 142 & 336 & GPCR, neuron morphogenesis & nerve ring, ventral cord neuron, AIY \\
{\em ceh-36} & 25 & 58 & cation symporter, myosin complex & pharynx, germ line, intestine \\
{\em ceh-6} & 105 & 226 & GPCR, axon guidance & ventral cord neuron, nerve ring \\
{\em cnd-1} & 22 & 394 & regulation of transcription, oviposition & QR.a/p, Z1.aa, C(a/p)p \\
{\em F21D5.9} & 545 & 495 & substrate-specific channel, oxygen transport & nerve ring, head/tail neuron \\
{\em hlh-16} & 544 & 1309 & signal transduction, chemosensory & amphid, phasmid neuron \\
{\em irx-1} & 102 & 213 & post-synaptic membrane, unfolded protein response & body wall muscle, ASE, PVN \\
{\em mir-57} & 583 & 213 & oxidoreductase activity, carboxylic acid metabolism & hypodermis, intestine, seam cell, germ line \\
{\em mls-2} & 140 & 397 & signal transduction, nerve terminal & AFD, ASK, CAN, ASI, AIY, ASG neurons \\
{\em pal-1} & 213 & 887 & Wnt ligand ({\em cwn-1, mom-2, egl-20, lin-44}), oviposition & gubernacular muscles, Z1.pp      \\
{\em pha-4} & 72 & 567 & protease inhibitor, UPR, lipid metabolism & pharynx, germ line, intestine \\
{\em ttx-3} & 133 & 169 & receptor activity, membrane & intestine, germ line, nervous system \\
{\em unc-130} & 146 & 393 & membrane, GPCR & preanal ganglion, excretory cell \\
\hline
{\em ceh-36} vs. - & 158 & 118 & & \\
{\em cnd-1} vs. - & 42 & 602 & & \\
{\em pha-4} vs. - & 482 & 714 & &  \\
\end{tabular}
\end{table}

\end{frame}

\subsection{Alternative splicing}

\begin{frame}{{\em alp-1} appears to express a different splice form in the pharynx}
\begin{center}
\includegraphics[height=0.8\textheight]{alp-1_1.png}
\end{center} 
\end{frame}

\subsection{Antisense transcription}

\begin{frame}{Antisense transcription enriched in specific tissues: {\em nhr-236}}

\includegraphics[width=1\textwidth]{nhr-236_2.png}

\end{frame}

\begin{frame}{A tissue-specific bidirectional promoter}

\includegraphics[width=1\textwidth]{C25E10_13.png}

\end{frame}

\begin{frame}{Transcriptomes of cells expressing a marker}

\begin{itemize}

\item Many genes are enriched or depleted in samples expressing a
particular TF.

\item The sets of genes enriched or depleted are often consistent with
known annotation of expression patterns.

\item More genes are enriched relative to the corresponding negative
sort sample, than compared to one control.

\item We can see differences in alternative splicing and antisense
transcription between tissues.

\end{itemize}

\end{frame}

\section{Genes enriched in cells expressing several markers}

\subsection{Hierarchical clustering}

\begin{frame}{Clusters of genes enriched in the same sort fractions}
\begin{minipage}{0.5\textwidth}
\includegraphics[width=\columnwidth]{TMEV4.png}
\end{minipage}
\begin{minipage}{0.48\textwidth}
{\small 
\begin{itemize}

\item {\em che-11} and {\em ifta-1} are localized to cilia in
amphid and phasmid neurons.

\item Presumably co-clustered genes are expressed
in similar cells.

\end{itemize}
}
\end{minipage}
\vspace{5mm}

\hfill TMEV (TIGR Multi-Experiment Viewer)
\end{frame}

\begin{frame}{Clusters of genes enriched in the same sort fractions}

\begin{minipage}{0.45\textwidth}
\includegraphics[width=0.9\columnwidth]{TMEV5.png}
\end{minipage}
\begin{minipage}{0.48\textwidth}
\begin{itemize}

\item {\em unc-96, unc-45, ketn-1} are expressed in muscle ({\em ketn-1}
also is expressed in pharynx)

\item Presumably, co-clustering genes also are expressed there.
\end{itemize}

\end{minipage}

\hfill TMEV (TIGR Multi-Experiment Viewer)
\end{frame}

\begin{frame}{Genes enriched in cells expressing several TFs}

\begin{itemize}

\item Many clusters of genes are coexpressed across the sorted fractions,
suggesting that they are expressed in similar tissues.

\item Unmixing is not currently working well. Including negative fractions,
and correcting for sort purity, may help.

\end{itemize}

\end{frame}

\section{Summary}

\begin{frame}{Summary}

\begin{itemize}

\item RNA-seq finds thousands of genes enriched in particular sets
of cells.
\pause

\item These enrichments are generally consistent with
expression of many known genes.
\pause

\item The combined expression data
set is informative about tissue-specific expression.
\pause

\item We can see examples of splicing and antisense transcription
which are enriched in different tissues.

\end{itemize}

\end{frame}

\begin{frame}{Cell-type-specific expression data can suggest regulatory relationships}

\includegraphics[height=0.8\textheight]{imported/novoshtern2011fig3.jpg}

Human hematopoeisis (Novoshtern et al., 2011)

{\small flow sorted blood cells into groups, then profiled expression of each group}

\end{frame}

\begin{frame}{Future directions}

\begin{itemize}

\item Quantify differences in splicing and antisense transcription between
sorted fractions
\pause

\item Predict regulatory targets of transcription factors, by combining
expression data with motif, chromatin and transcription factor ChIP data
\pause

\end{itemize}

\end{frame}

\begin{frame}{Thank you}

\begin{itemize}

\item John Murray, and the Murray lab (particularly for lineaging)

\item particularly Travis Walton, who did all of the worm prep, FACS sorting,
and library preparation

\item sequencing core

\item many others, for useful questions and commments

\end{itemize}

\end{frame}


\end{document}

