\documentclass[serif,9pt]{beamer}
\usepackage[latin1]{inputenc}
\usepackage{multirow}
\renewcommand{\sfdefault}{times}

% \usetheme{Warsaw}

\graphicspath{
{/home/jburdick/gcb/work/svnroot/trunk/tex/image/}
{/home/jburdick/gcb/work/svnroot/trunk/tex/thesis/proposal/image/}
{/home/jburdick/gcb/work/svnroot/trunk/src/}
{/home/jburdick/gcb/}
}
\beamertemplatenavigationsymbolsempty

\title{Where are genes expressed in the {\em C. elegans} embryo?}
\author{Josh Burdick \\ Murray Lab}
\date{September 13, 2012}
\begin{document}

\begin{frame}
\titlepage
\end{frame}

\begin{frame}{Where a gene is expressed is important}

\begin{centering}
\includegraphics[width=\textwidth]{notch_example_1.png}
\end{centering}

\hfill Neves and Priess, 2005

\end{frame}

\begin{frame}{Where {\em many} genes are expressed is important}

\includegraphics[width=\textwidth]{yanai2008.png}

\hfill Yanai, 2008

\end{frame}

\begin{frame}{Measuring expression using automated lineage tracing}

{\center
\includegraphics[width=0.75\textwidth]{image/movie/P0_1200x400/20110929_ceh-43_L1.png}

\includegraphics[width=0.75\textwidth]{image/movie/P0_1200x400/20101213_RW10889_pal-1_L3.png}

\includegraphics[width=0.75\textwidth]{image/movie/P0_1200x400/20110531_pha4_24C_L3.png}
}

\end{frame}

\begin{frame}{Where and when is each gene expressed in development?}

\begin{itemize}

\item What genes are enriched in cells that express some TF?

\pause 

\item What genes are enriched in cells expressing several TFs?

\pause

\item What does this imply about how genes are regulated?

\end{itemize}

\end{frame}

\begin{frame}{Samples sorted}

\includegraphics[width=\textwidth]{FACSlog.png}

\hfill Travis Walton

{\small promoter fusions are mCherry, protein fusions are GFP}

\end{frame}

\begin{frame}{How reproducible are the data?}

\includegraphics[width=0.8\textwidth]{git/unmix/seq/quant/replicateScatterPlot.png}

\end{frame}

\begin{frame}{Which samples are similar to each other?}

\includegraphics[width=1\textwidth]{git/unmix/seq/cluster/pvclust.pdf}

\end{frame}

\begin{frame}{Are some cells depleted from the sorting process?}

\includegraphics[width=0.8\textwidth]{git/unmix/seq/FACS/ungated_and_singlets.pdf}

\end{frame}


\begin{frame}{Some genes are mostly expressed at a particular time}

\includegraphics[height=0.8\textheight]{git/unmix/seq/timing/gene_timeseries.pdf}

\hfill Boeck and Waterston {\em et al}, 2012

\end{frame}

\begin{frame}{Picking time-specific genes}

\includegraphics[width=\textwidth]{time_figure.png}

\hfill Boeck and Waterston {\em et al}, 2012

\end{frame}

\begin{frame}{What times are included in the sorting?}

\begin{centering}
\includegraphics[width=\textwidth]{git/unmix/seq/timing/perStageMarkersUngated.pdf}
\end{centering}

\end{frame}

\begin{frame}{{\em cnd-1}}
\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{R/unmix/sort_paper/seq/fraction/plots_and_stats/cnd-1.pdf}
\end{minipage}
\begin{minipage}{0.58\textwidth}
\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
germ line & 295 & $10^{-17}$ \\
QR.a/p & 8 & 0.005 \\
cc(A/P)(L/R) & 8 & 0.005 \\
D & 25 & 0.0001 \\
C(a/p)p & 25 & 0.0002 
\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline

\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20110731_RW10434_L1.png}

\end{frame}

\begin{frame}{{\em unc-130}}

\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{R/unmix/sort_paper/seq/fraction/plots_and_stats/unc-130.pdf}
\end{minipage}
\begin{minipage}{0.58\textwidth}
\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline

\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline

\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20110927_RW11144_L4.png}

\end{frame}

\begin{frame}{{\em ceh-6}}

\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{R/unmix/sort_paper/seq/fraction/plots_and_stats/ceh-6.pdf}
\end{minipage}
\begin{minipage}{0.58\textwidth}
\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
excretory secretory system & 5 & 0.006 
\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline
G-protein coupled receptor signalling & {\em srw-113, gnrr-7, srg-31} (12 total) & $10^{-4}$ \\
axon guidance & {\em ceh-17, zag-1, unc-6} & $10^{-3}$ \\
\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20120331_RW10871_L4.png}

\end{frame}

\begin{frame}{{\em ceh-26}}

\begin{minipage}{0.4\textwidth}
\includegraphics[width=\textwidth]{R/unmix/sort_paper/seq/fraction/plots_and_stats/ceh-26.pdf}
\end{minipage}
\begin{minipage}{0.58\textwidth}
\begin{table}[!tbp]\scriptsize
\begin{tabular}{lll}
anatomy term & \# genes & adjusted $p$ \\
\hline
excretory secretory system & 5 & 0.006 
\end{tabular}
\end{table}
\end{minipage}

\begin{table}\footnotesize
\begin{tabular}{llr}
GO category & genes & $p$ value \\
\hline

\end{tabular}
\end{table}

\includegraphics[width=\textwidth]{image/movie/P0_1200x400/20120401_JIM122_L3.png}

\end{frame}

\begin{frame}{Genes enriched in sorted fractions}

\begin{table}\footnotesize
\begin{tabular}{lllll}
sort & \# genes & \# genes & enriched GO & enriched \\
marker & enriched & depleted & categories & tissues \\
\hline
\end{tabular}
\end{table}

\end{frame}


\begin{frame}{Clusters of genes enriched in the same sort fractions}

FIXME


\end{frame}

\begin{frame}{Fourteen reporters distinguish many cells}

\begin{figure}
\begin{centering}
\includegraphics[width=0.9\textwidth]{R/unmix/sort_paper/plot/distinguishableCells.pdf}

\end{centering}
\end{figure}

\end{frame}

\begin{frame}{Cell-type-specific expression data can suggest regulatory relationships}

\includegraphics[height=0.8\textheight]{imported/novoshtern2011fig3.jpg}

Human hematopoeisis (Novoshtern et al., 2011)

{\small flow sorted blood cells into groups, then profiled expression of each group}

\end{frame}


\begin{frame}{Antisense transcription enriched in specific tissues: {\em nhr-236}}

\includegraphics[width=1\textwidth]{IGV/nhr-236_2.png}

\end{frame}

\begin{frame}{A tissue-specific bidirectional promoter}

\includegraphics[width=1\textwidth]{IGV/C25E10_13.png}

\end{frame}

\begin{frame}{{\em alp-1} appears to express a different splice form in the pharynx}
\begin{center}
\includegraphics[height=0.8\textheight]{IGV/alp-1_1.png}
\end{center} 
\end{frame}

\begin{frame}{Summary}

\begin{itemize}

\item RNA-seq of sorted samples is consistent with
expression of many known genes.

\item Although unmixing isn't working, the combined expression data
set is informative about tissue-specific expression.

\item We can see examples of splicing and antisense transcription
which are enriched in different tissues.

\end{itemize}

\end{frame}

\begin{frame}{Future directions}

\begin{itemize}

\item Apply prediction method to estimate where genes are expressed.

\item Measure accuracy of predictions, using RNA-FISH and fluorescent reporters.

\item Predict regulatory targets of transcription factors, by combining
expression data with motif, chromatin and transcription factor ChIP data.

\item Determine significance of particular genes being expressed in particular cells.

\end{itemize}

\end{frame}



\begin{frame}{Thank you}

\begin{itemize}

\item John Murray, and the Murray lab (particularly for lineaging)

\item particularly Travis Walton, who did all of the worm prep, FACS sorting,
and library preparation

\item sequencing core

\item many others, for useful questions and commments

\end{itemize}

\end{frame}


\end{document}

