\documentclass{article}
\usepackage{amsfonts}
\usepackage[cm]{fullpage}
\usepackage{graphicx}
\usepackage{hyperref}

\begin{document}

\section*{Comparing sorted expression with temporal data}

In constructing the cell-sorting matrix, we need an estimate of which cells
were actually included in the sorting. Earlier cells, being larger, presumably
are less likely to be included.

Therefore, we compare the FACS-sorted data with a timecourse of expression
in embryogenesis.

\subsection*{Normalization}

It is not obvious how to normalize RNA-seq data in a biologically useful way.
Presumably, if $r$ out of $N$ reads map to a gene, then the ratio $r/N$ is
proportional to the amount of RNA from that transcript. The number of copies
of an individual transcript per cell also depends on the mixture of splice forms
being expressed, and on the total amount of RNA per cell (which in turn
depends on cell size, among other things.)

Robinson and Oshlack suggest using a robust mean of log-transformed read counts
in order to normalize between RNA-seq experiments; this is called TMM-normalization.
To see if this approach is
relevant here, we plot log-transformed expression of pairs of timepoints.
Some examples are shown below.

\includegraphics[width=\textwidth]{logRatioPlots1.pdf}

For most pairs of genes, there is a subset of genes that doesn't change between
timepoints, and this subset appears larger for adjacent timepoints; that subset is
indeed slightly skewed between samples (for instance, between times 690 and 720.)
However, many genes are switching on and off between more distant timepoints,
which suggests
that TMM-normalization might require modification for this dataset.

\subsection*{Determining time-specific genes}

Genes in the developing embryo have many patterns (cite Baugh paper.)
For the purpose of determining which time points are included in our FACS-sorted
data, a useful subset of genes are those which are on only at a specific time.

To determine such a set, we fit each gene's (log transformed) expression
profile to a normal curve.
For now, we use the normalized coverage within a gene's bounds as a measure
of expression, with the caveat that we are measuring proportion of expression per
sample (which only corresponds approximately to transcript abundance.)
We focus on the 8,379 genes which had an average coverage $\ge 10 ppm$ here.
If $x_t$ is the expression of a gene at time $t$,
we define a weight

\[
w_t = \frac{x_t}{\sum_t x_t}
\]
and find the ``mean time expressed'', $\mu$:

\[
\mu_t = \sum_{t = 0,30,60,...} w_t \cdot t
\]

We define the ``standard deviation of time expressed'' $\sigma_t$ analogously.


\begin{center}
\includegraphics[width=0.6\textwidth]{timeMeanAndSD.pdf}
\end{center}

Genes with a small $\sigma_t$ are mostly expressed at one time; genes with
other temporal profiles will have a larger $\sigma_t$. We use the 1,676
genes with the lowest 20\% of standard deviation ($\sigma_t < 137$)
as ``time-specific marker genes.''
This includes genes such as {\em tbx-37} ($\mu_t = 83$) and {\em fkh-3}
($\mu_t = 84$.)

\subsection*{Genes expressed in FACS-sorted cells}

If cells from a particular time are present in a sorted fraction, then a
gene specific to that time may be included (if, at that time, that gene was present in
that sorted fraction.)
If we don't see any markers from some time, however,
that suggests that cells from that time weren't included in the sorting.

\includegraphics[width=\textwidth]{perStageMarkersInSorted.pdf}

A few of the very early time-specific genes were expressed in the sorted samples.
However, it seems that more genes from later (say, after $\mu_t = 100$)
were expressed. This suggests that we are mostly including cells after this time.

\end{document}


